client/src/main/java/org/eclipse/hono/client/impl/ProtocolAdapterCommandConsumerFactoryImpl.java

```diff
@@ -0,0 +1,332 @@
+import org.eclipse.hono.client.CommandTargetMapper;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.MessageConsumer;
+import org.eclipse.hono.client.ProtocolAdapterCommandConsumerFactory;
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.ResourceIdentifier;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.eventbus.Message;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+
+/**
+ * A factory for creating clients for the <em>AMQP 1.0 Messaging Network</em> to receive commands and send responses.
+ * <p>
+ * The factory uses two kinds of consumer links to receive commands:
+ * <ul>
+ * <li>A single consumer link on an address containing the protocol adapter instance id.</li>
+ * <li>A tenant-scoped link, created (if not already existing for that tenant) when
+ * {@link #createCommandConsumer(String, String, Handler)} is invoked.</li>
+ * </ul>
+ * <p>
+ * Command messages are first received on the tenant-scoped consumer address. It is then determined whether there is
+ * a consumer and corresponding command handler for the command message's target device or one of the device's
+ * possible gateways. If found, that handler is either invoked directly, or, if it is on another protocol adapter
+ * instance, the command message is sent to that protocol adapter instance to be handled there.
+ */
+public class ProtocolAdapterCommandConsumerFactoryImpl extends AbstractHonoClientFactory implements ProtocolAdapterCommandConsumerFactory {
+
+    private static final int RECREATE_CONSUMERS_DELAY = 20;
+
+    /**
+     * Cache key used here is the tenant id.
+     */
+    private CachingClientFactory<MessageConsumer> mappingAndDelegatingCommandConsumerFactory;
+
+    private final String adapterInstanceId;
+    private final AdapterInstanceCommandHandler adapterInstanceCommandHandler;
+    private final AtomicBoolean recreatingConsumers = new AtomicBoolean(false);
+    private final AtomicBoolean tryAgainRecreatingConsumers = new AtomicBoolean(false);
+
+    private BasicDeviceConnectionClientFactory deviceConnectionClientFactory;
+    private MappingAndDelegatingCommandHandler mappingAndDelegatingCommandHandler;
+    private ProtonReceiver adapterSpecificConsumer;
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+
+    /**
+     * Creates a new factory for an existing connection.
+     * 
+     * @param connection The connection to the AMQP network.
+     * @param adapterInstanceId The id of the protocol adapter instance that this factory is running in.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public ProtocolAdapterCommandConsumerFactoryImpl(final HonoConnection connection, final String adapterInstanceId) {
```

Comment:
I had hoped that we would be able to get rid of the second parameter altogether. But that might be done in a future PR ...